package NKikimrLongTxService;

option java_package = "ru.yandex.kikimr.proto";

import "ydb/public/api/protos/ydb_status_codes.proto";
import "ydb/public/api/protos/ydb_issue_message.proto";

message TLongTxId {
    optional bytes UniqueId = 1;
    optional uint32 NodeId = 2;
    optional uint64 SnapshotStep = 3;
    optional uint64 SnapshotTxId = 4;
}

message TEvBeginTx {
    enum EMode {
        MODE_UNSPECIFIED = 0;
        MODE_READ_ONLY = 1;
        MODE_WRITE_ONLY = 2;
        MODE_READ_WRITE = 3;
    }

    optional string DatabaseName = 1;
    optional EMode Mode = 2;
}

message TEvBeginTxResult {
    optional Ydb.StatusIds.StatusCode Status = 1;
    repeated Ydb.Issue.IssueMessage Issues = 2;
    optional TLongTxId LongTxId = 3;
}

message TEvCommitTx {
    optional TLongTxId LongTxId = 1;
}

message TEvCommitTxResult {
    optional Ydb.StatusIds.StatusCode Status = 1;
    repeated Ydb.Issue.IssueMessage Issues = 2;
}

message TEvRollbackTx {
    optional TLongTxId LongTxId = 1;
}

message TEvRollbackTxResult {
    optional Ydb.StatusIds.StatusCode Status = 1;
    repeated Ydb.Issue.IssueMessage Issues = 2;
}

message TEvAttachColumnShardWrites {
    message TWrite {
        optional uint64 ColumnShard = 1;
        optional uint64 WriteId = 2;
    }

    optional TLongTxId LongTxId = 1;
    repeated TWrite Writes = 2;
}

message TEvAttachColumnShardWritesResult {
    optional Ydb.StatusIds.StatusCode Status = 1;
    repeated Ydb.Issue.IssueMessage Issues = 2;
}

message TEvAcquireReadSnapshot {
    optional string DatabaseName = 1;
}

message TEvAcquireReadSnapshotResult {
    optional Ydb.StatusIds.StatusCode Status = 1;
    repeated Ydb.Issue.IssueMessage Issues = 2;
    optional uint64 SnapshotStep = 3;
    optional uint64 SnapshotTxId = 4;
    optional string DatabaseName = 5;
}

message TEvSubscribeLock {
    optional uint64 LockId = 1;
    optional uint32 LockNode = 2;
}

message TEvLockStatus {
    enum EStatus {
        STATUS_UNSPECIFIED = 0;
        STATUS_SUBSCRIBED = 1;
        STATUS_NOT_FOUND = 2;
        STATUS_UNAVAILABLE = 3;
    }

    optional uint64 LockId = 1;
    optional uint32 LockNode = 2;
    optional EStatus Status = 3;
}

message TEvUnsubscribeLock {
    optional uint64 LockId = 1;
    optional uint32 LockNode = 2;
}
